### PR TITLE
Expose capability to invalidate cache based in prefix

### PR DESCRIPTION
**Pull request by AdamYahid**
This PR is building on the premise that when invalidating a directory in the file tree it's entire subtree should also be considered invalid, which currently is not the case

### CODE REVIEW

File: packages/cached/src/cached-fs.ts [hunk 1]
```diff
@@ -32,9 +32,17 @@ export function createCachedFs(fs: IFileSystem): ICachedFileSystem {
   const realpathCache = new Map<string, string>();
   const { promises } = fs;
   const invalidateAbsolute = (absolutePath: string) => {
-    const cachePath = getCanonicalPath(absolutePath);
-    realpathCache.delete(cachePath);
-    statsCache.delete(cachePath);
+    const cachePathPrefix = getCanonicalPath(absolutePath);
+    for (const key of realpathCache.keys()) {
```
Comment (AdamYahid) [inline packages/cached/src/cached-fs.ts:36]: in cases where we change x, x.y would also be invalidated, we need to find a way to make sure we don't invalidate things that shouldn't be invalidated

**Comment by AdamYahid** [comment]
The main concern with this PR is that it is very aggressive with cache invalidation and the potential issues that can arise are
1. The current issue that exists in the PR, which would invalidate files that shouldn't be invalided when invalidating a file with an additional suffix (for example invalidating a would invalid a.ts).
2. cache isn't aware of the concept of file or dir, it's just an entry in the cache - who is responsible for determining that. stating the path that is being invalided doesn't make sense (we wouldn't want to perform an fs action within an fs action), so how else?

**Comment by AdamYahid** [comment]
Given that cache isn't aware of files/directories, and therefor cannot make the distinction between deep/flat invalidation on it's own, a possible solution would be to the the consumer - either via invalidate API or via the actions they perform to decide wither to invalidate "all keys that start with" or just that specific key. 
The current state of the draft represents that mindset, and currently exposes said capability via the "invalidate" API with a new param(deep) which indicates that the consumer knows whether it's a directory or not, and if they need deep invalidation or not

There are several actions which can also benefit this distinction, which we can update to use this new capability such as
rmdir/rmdirsync
but there are some less obvious cases such as
rename, symlink and unlink (and all their variants) which require a bit more thought

File: packages/cached/test/cached-fs.spec.ts [hunk 2]
```diff
@@ -233,6 +233,31 @@ describe('createCachedFs', () => {
       expect(promiseStatSpy.callCount).to.equal(1);
       expect(stats).to.not.equal(stats2);
     });
+
+    it('deep invalidation should invalidate all the keys starting with the invalidation path', async () => {
+      const dirPath = '/dir';
+      const fileName = 'file';
+      const filePath = `${dirPath}/${fileName}`;
+      const memFs = createMemoryFs({ [filePath]: SAMPLE_CONTENT });
```
Comment (AviVahl) [inline packages/cached/test/cached-fs.spec.ts:241]:
```ts
createMemoryFs({
   dir: { file: SAMPLE_CONTENT },
   dir2: { file: SAMPLE_CONTENT }
});
```
and make sure invalidating `/dir` does not invalidate `/dir2/file`

File: packages/cached/src/cached-fs.ts [hunk 3]
```diff
@@ -181,8 +194,12 @@ export function createCachedFs(fs: IFileSystem): ICachedFileSystem {
         }
       },
     }),
-    invalidate(path) {
-      return invalidateAbsolute(fs.resolve(path));
+    invalidate(path, deep = false) {
+      if (deep) {
```
Comment (AviVahl) [inline packages/cached/src/cached-fs.ts:198]: single line with a trinary

**Pull request merged**